File: Spec/lib/budgets.ts
import { and, eq, gte, lt, sql } from "drizzle-orm";
import { db } from "@/lib/db";
import { budgets, transactions, type Budget } from "@/db/schema";

export type Period = "week" | "month" | "year";
export interface PeriodRange {
  start: Date;
  end: Date;
}

function parseAnchor(startsOn: string | Date): Date {
  if (startsOn instanceof Date) return startsOn;
  const d = new Date(startsOn + "T00:00:00Z");
  return d;
}

function addPeriods(anchor: Date, period: Period, count: number): Date {
  const d = new Date(anchor);
  if (period === "week") {
    d.setUTCDate(d.getUTCDate() + count * 7);
  } else if (period === "month") {
    d.setUTCMonth(d.getUTCMonth() + count);
  } else {
    d.setUTCFullYear(d.getUTCFullYear() + count);
  }
  return d;
}

export function periodAt(
  startsOn: string | Date,
  period: Period,
  at: Date,
): PeriodRange {
  const anchor = parseAnchor(startsOn);
  if (at < anchor) {
    return { start: anchor, end: addPeriods(anchor, period, 1) };
  }
  let lo = 0;
  let hi = 1;
  while (addPeriods(anchor, period, hi) <= at) {
    lo = hi;
    hi *= 2;
    if (hi > 10_000) break;
  }
  while (lo + 1 < hi) {
    const mid = Math.floor((lo + hi) / 2);
    if (addPeriods(anchor, period, mid) <= at) lo = mid;
    else hi = mid;
  }
  return {
    start: addPeriods(anchor, period, lo),
    end: addPeriods(anchor, period, lo + 1),
  };
}

export interface BudgetProgress {
  budget: Budget;
  period: PeriodRange;
  spentEur: number;
}

export async function budgetProgress(budget: Budget, at: Date = new Date()): Promise<BudgetProgress> {
  const range = periodAt(budget.startsOn, budget.period as Period, at);
  const [{ total }] = await db
    .select({
      total: sql<string>`coalesce(sum(${transactions.amountEur}), 0)`,
    })
    .from(transactions)
    .where(
      and(
        eq(transactions.categoryId, budget.categoryId),
        eq(transactions.isTransfer, false),
        eq(transactions.direction, "debit"),
        gte(transactions.bookedAt, range.start),
        lt(transactions.bookedAt, range.end),
      ),
    );
  const spentEur = Math.abs(Number(total ?? "0"));
  return { budget, period: range, spentEur };
}

export async function activeBudgetsProgress(at: Date = new Date()): Promise<BudgetProgress[]> {
  const rows = await db.select().from(budgets).where(eq(budgets.active, true));
  return Promise.all(rows.map((b) => budgetProgress(b, at)));
}

File: Core/Sources/CoreLogic/Budgets.swift
import Foundation
import SwiftData
import CoreModel

extension CoreLogic {
    // Ports lib/budgets.ts. A budget recurs from `startsOn` by its period; `periodAt`
    // finds the active window containing `at`, and progress is the abs sum of debit
    // amountEur in that window for the budget's category (transfers excluded).
    public enum Budgets {
        public struct PeriodRange: Equatable, Sendable {
            public let start: Date
            public let end: Date
        }

        public struct Progress: Equatable, Sendable {
            public let budgetId: UUID
            public let categoryId: UUID?
            public let amountEur: Decimal
            public let period: BudgetPeriod
            public let range: PeriodRange
            public let spentEur: Decimal
        }

        public static func addPeriods(_ anchor: Date, _ period: BudgetPeriod, _ count: Int) -> Date {
            var cal = Calendar(identifier: .iso8601)
            cal.timeZone = TimeZone(identifier: "UTC")!
            switch period {
            case .week:
                return cal.date(byAdding: .day, value: count * 7, to: anchor) ?? anchor
            case .month:
                return cal.date(byAdding: .month, value: count, to: anchor) ?? anchor
            case .year:
                return cal.date(byAdding: .year, value: count, to: anchor) ?? anchor
            }
        }

        public static func periodAt(startsOn anchor: Date, period: BudgetPeriod, at: Date) -> PeriodRange {
            if at < anchor {
                return PeriodRange(start: anchor, end: addPeriods(anchor, period, 1))
            }
            var lo = 0
            var hi = 1
            while addPeriods(anchor, period, hi) <= at {
                lo = hi
                hi *= 2
                if hi > 10_000 { break }
            }
            while lo + 1 < hi {
                let mid = (lo + hi) / 2
                if addPeriods(anchor, period, mid) <= at { lo = mid } else { hi = mid }
            }
            return PeriodRange(
                start: addPeriods(anchor, period, lo),
                end: addPeriods(anchor, period, lo + 1)
            )
        }

        @MainActor
        public static func budgetProgress(
            _ budget: Budget,
            at: Date,
            in ctx: ModelContext
        ) throws -> Progress {
            let range = periodAt(startsOn: budget.startsOn, period: budget.period, at: at)
            var spent: Decimal = 0
            if let catId = budget.category?.id {
                let start = range.start
                let end = range.end
                let txs = try ctx.fetch(FetchDescriptor<Transaction>(
                    predicate: #Predicate { tx in
                        tx.category?.id == catId &&
                        tx.isTransfer == false &&
                        tx.bookedAt >= start &&
                        tx.bookedAt < end
                    }
                ))
                for tx in txs where tx.direction == .debit {
                    spent += tx.amountEur ?? 0
                }
            }
            return Progress(
                budgetId: budget.id,
                categoryId: budget.category?.id,
                amountEur: budget.amountEur,
                period: budget.period,
                range: range,
                spentEur: abs(spent)
            )
        }

        @MainActor
        public static func activeBudgetsProgress(
            at: Date,
            in ctx: ModelContext
        ) throws -> [Progress] {
            let budgets = try ctx.fetch(FetchDescriptor<Budget>(
                predicate: #Predicate { $0.active == true }
            ))
            return try budgets.map { try budgetProgress($0, at: at, in: ctx) }
        }
    }
}
